core/common/scale.cpp

```diff
@@ -0,0 +1,506 @@
+
+      expected::Result<Tristate, TristateDecodeError> decodeTristate(
+          uint8_t byte) {
+        switch (byte) {
+          case 0:
+            return expected::Value{Tristate::kNone};
+          case 1:
+            return expected::Value{Tristate::kFalse};
+          case 2:
+            return expected::Value{Tristate::kTrue};
+          default:
+            break;
+        }
+
+        return expected::Error{TristateDecodeError::kUnexpectedValue};
+      }
+
+      auto encodeInt8(int8_t value) {
+        return value;
+      }
+
+      auto encodeUInt8(uint8_t value) {
+        return value;
+      }
+
+      auto encodeInt16(uint16_t value) {
+        return encodePlainInteger(value);
+      }
+
+      auto encodeUint16(uint16_t value) {
+        return encodePlainInteger(value);
+      }
+
+      auto encodeInt32(uint32_t value) {
+        return encodePlainInteger(value);
+      }
+
+      auto encodeUint32(uint32_t value) {
+        return encodePlainInteger(value);
+      }
+
+      auto encodeInt64(int64_t value) {
+        return encodePlainInteger(value);
+      }
+
+      auto encodeUint64(uint64_t value) {
+        return encodePlainInteger(value);
+      }
+
+      auto decodeInt8(int8_t value) {
+        return value;
+      }
+
+      auto decodeUInt8(uint8_t value) {
+        return value;
+      }
+
+      auto decodeInt16(uint16_t value) {
+        return decodePlainInteger(value);
+      }
+
+      auto decodeUint16(uint16_t value) {
+        return decodePlainInteger(value);
+      }
+
+      auto decodeInt32(uint32_t value) {
+        return decodePlainInteger(value);
+      }
+
+      auto decodeUint32(uint32_t value) {
+        return decodePlainInteger(value);
+      }
+
+      auto decodeInt64(int64_t value) {
+        return decodePlainInteger(value);
+      }
+
+      auto decodeUint64(uint64_t value) {
+        return decodePlainInteger(value);
+      }
+
+      uint8_t encodeBool(bool value) {
+        return value ? static_cast<uint8_t>(0x01) : static_cast<uint8_t>(0x00);
+      }
+
+      DecodeBoolResult decodeBool(uint8_t byte) {
+        if (byte == 0) {
+          return expected::Value{false};
+        } else if (byte == 1) {
+          return expected::Value{true};
+        }
+        return expected::Error{DecodeBoolError::kUnexpectedValue};
+      }
+    }  // namespace fixedwidth
+
+    namespace compact {
+      struct EncodingCategoryLimits {
+        enum type : size_t {
+          kLessThan64 = 1ul << 6,
+          kLessThan16384 = 1ul << 14,
+          kLessThanTwoDegreeThirty = 1ul << 30
+        };
+      };
+
+      template <class T>
+      struct EncodedLengthCalculator {
+        EncodedLengthCalculator() {}
+
+        size_t operator()(const T &) {
+          assert(false && "not implemented for this type");
+          return 0;
+        }
+      };
+
+      EncodeIntegerResult encodeUint8(uint8_t value) noexcept {
+        if (value < EncodingCategoryLimits::kLessThan64) {
```

Comment:
converted to text: "if value less than kLessThan64 then ...", ain't it sounds weird for you?

Reply:
Self-describing on the other hand. Any proposals?

Reply:
Ok, renamed:
struct EncodingCategoryLimits {
enum type : size_t {
kMinUint16 = 1ul << 6,
kMinUint32 = 1ul << 14,
kMinBigInteger = 1ul << 30
};
};

OK?

Reply:
imo looks better, yeah

Reply:
Great.